neo4j/resultsummary.go

```diff
@@ -228,13 +247,73 @@ type Notification interface {
 	// In that case, Category returns UnknownCategory while RawCategory returns the raw string
 	RawCategory() string
 	// SeverityLevel returns the mapped security level of this notification.
-	// If the severity level is not a known value, SeverityLevel returns UnknownSeverity
+	// If the severity level is not a known value, SeverityLevel returns notifications.UnknownSeverity
 	// Call RawSeverityLevel to get access to the raw string value
-	SeverityLevel() NotificationSeverity
+	SeverityLevel() notifications.NotificationSeverity
 	// Category returns the mapped category of this notification.
 	// If the category is not a known value, Category returns UnknownCategory
 	// Call RawCategory to get access to the raw string value
-	Category() NotificationCategory
+	//lint:ignore SA1019 NotificationCategory is supported at least until 6.0
+	Category() notifications.NotificationCategory
+}
+
+// GqlStatusObject represents a GqlStatusObject generated when executing a statement.
+// A GqlStatusObject can be visualized in a client pinpointing problems or other information about the statement.
+// Contrary to failures or errors, GqlStatusObjects do not affect the execution of the statement.
+//
+// GqlStatusObject is part of the GQL compliant notifications preview feature
+// (see README on what it means in terms of support and compatibility guarantees)
+type GqlStatusObject interface {
```

Comment:
Pretty much just the pattern I followed for most of the Go driver API:

`Notification` struct in summary.go -> `Notification` interface in resultsummary.go
`Plan` struct in summary.go -> `Plan` interface in resultsummary.go
`ProfiledPlan` struct in summary.go -> `ProfiledPlan` interface in resultsummary.go
`InputPosition` struct in summary.go -> `InputPosition` interface in resultsummary.go

etc etc.

I think it is done like this as most of the driver API surface returns interfaces rather than raw structs to let us have some flexibility on updating the implementation details of the structs without breaking the exposed interfaces (this ties in with the above conversation of having Code and Title in the GqlStatusObject struct and not the GqlStatusObject interface).

Reply:
leaving open for discussion

Reply:
As discussed this morning, ideally, we might be able to move most of these public structs to the internal package and only export the interface counterparts. Currently gives lots of cyclic dependency issues/exposes internal structs to the client. Needs more thought I've carded this up for future improvement to driver.